datasource/ethereum/src/block_ingestor.rs

```diff
@@ -0,0 +1,178 @@
+            logger: logger.new(o!("component" => "BlockIngestor")),
+            polling_interval,
+        })
+    }
+
+    pub fn into_polling_stream(self) -> impl Future<Item = (), Error = ()> {
+        // Currently, there is no way to stop block ingestion, so just leak self
+        let static_self: &'static _ = Box::leak(Box::new(self));
+
+        // Create stream that emits at polling interval
+        tokio::timer::Interval::new(Instant::now(), static_self.polling_interval)
+            .map_err(move |e| {
+                error!(static_self.logger, "timer::Interval failed: {:?}", e);
+            })
+            .for_each(move |_| {
+                // Attempt to poll
+                static_self.do_poll().then(move |result| {
+                    if let Err(e) = result {
+                        // Some polls will fail due to transient issues
+                        warn!(
+                            static_self.logger,
+                            "failed to poll for latest block: {:?}", e
+                        );
+                    }
+
+                    // Continue polling even if polling failed
+                    future::ok(())
+                })
+            })
+    }
+
+    fn do_poll<'a>(&'a self) -> impl Future<Item = (), Error = Error> + 'a {
+        // Ask for latest block from Ethereum node
+        self.get_latest_block()
+            .and_then(move |latest_block: Block<Transaction>| {
+                // Store latest block in block store.
+                // Might be a no-op if latest block is one that we have seen.
+                // ingest_blocks will return a (potentially incomplete) list of blocks that are
+                // missing.
+                self.ingest_blocks(stream::once(Ok(latest_block)))
+            })
+            .and_then(move |missing_block_hashes| {
+                // Repeatedly fetch missing blocks, and ingest them.
+                // ingest_blocks will continue to tell us about more missing blocks until we have
+                // filled in all missing pieces of the blockchain (that we care about).
+                future::loop_fn(
+                    missing_block_hashes,
+                    move |missing_block_hashes| -> Box<Future<Item = _, Error = _> + Send> {
+                        if missing_block_hashes.is_empty() {
+                            // If no blocks were missing, then the block head pointer was updated
+                            // successfully, and this poll has completed.
+                            Box::new(future::ok(future::Loop::Break(())))
+                        } else {
+                            // Some blocks are missing: load them, ingest them, and repeat.
+                            let missing_blocks = self.get_blocks(&missing_block_hashes);
+                            Box::new(self.ingest_blocks(missing_blocks).map(
+                                |missing_block_hashes| future::Loop::Continue(missing_block_hashes),
```

Comment:
These new comments are great at describing what is being done but I still don't get the why. My previous doubt still stands, in what situation would this call to `ingest_blocks` update the head pointer? Can this loop forever or is there a maximum number of iterations?